### PR TITLE
Advent 2018: Splitting Data With CDC

### PR DESCRIPTION
**Pull request by fd0**
So, here's the first version of my post for the Advent 2018. Please let me know if there are any errors (typos or otherwise)! :)

### CODE REVIEW

File: content/advent-2018/split-data-with-cdc.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+if err != nil {
+	panic(err)
+}
+
+fmt.Printf("using polynomial %v for splitting data\n", p)
+
+chunker := chunker.New(os.Stdin, p)
+```
+
+By default, it tries to make chunks of one megabyte on average (you can control that with the method `SetAverageBits()`). Then we can repeatedly call the `Next()` function on the `Chunker` and pass in a byte slice buffer for the data read from the reader passed to `New()`, it'll return a `Chunk` type and an error:
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: > "By default, it tries to make chunks of one megabyte on average <..>

Do you think this is important for this paragraph? I feel that it can be omitted here, leaving a reader focused on what is happening in the example.

---

Can I suggest rephrasing the next sentence to something like

"The `Next()` *method* on the `Chunker` reads data out the reader into a byte slice buffer. The method returns a next `Chunk` and an error. We call the method untill it returns `io.EOF` error, to read all the data:"

File: content/advent-2018/split-data-with-cdc.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+8297953 589913  0004cf802b600000
+8887866 554526  001eb5a362900000
+9442392 1307416 000ef1e549c00000
+[...]
+102769045       864607  00181b67df300000
+103633652       592134  0000a0c8b4200000
+104225786       631814  001d5ba20d38998b
+```
+
+On a second run, it'll generate a new polynomial and the chunks will be different. Let's fix the polynomial (`0x3dea92648f6e83`) for the next runs and change the program like this ([code](https://gist.github.com/bf984e1c40b56eaeff310d07a0d71128)):
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: How about you move the **Important** note from the above to here? I think there's a strong connection between these two sentences since here you're "fixing" what has been described in the important part.

File: content/advent-2018/split-data-with-cdc.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+```go
+chunker := chunker.New(os.Stdin, 0x3dea92648f6e83)
+```
+
+Running the new program will give us the same chunks again:
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: Maybe

"After replacing a randomly generated polynomial to a constant every new run will give us the same chunks:"

File: content/advent-2018/split-data-with-cdc.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+```
+$ (echo -n foo; cat data) | go run main.go
+0       2908787 000c31839a100000        f41084d25bc273f2[...]
+2908787 2088949 001e8b4104900000        ba7c71226609c0de[...]
+4997736 1404824 000a18d1f0200000        9c4db21626a27d4b[...]
+6402560 617070  001bc6ac84300000        4c249c3b88500c23[...]
+7019630 1278326 001734984b000000        5bec562b7ad37b8b[...]
+```
+
+The SHA-256 hash for the first chunk changed from `8d17f5f7326a2fd7[...]` to `f41084d25bc273f2[...]`, but the rest is still the same. That's pretty cool.
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: > [..] but the rest **~~is~~** are still the same.

File: content/advent-2018/split-data-with-cdc.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+Let's see how many different chunks we have in our data file byte counting the different hashes:
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: > [..] in our data file **byte counting** the different hashes:

I think there's a typo here. Did you mean "by counting" the hashes?

File: content/advent-2018/split-data-with-cdc.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+Changing a few consecutive bytes of data somewhere in most cases also only affects a single chunk, let's write the output to a file, change the data a bit using `sed`, and write another file:
+
+```
+$ cat data | go run main.go > orig.log
+$ cat data | sed 's/EZE8HX/xxxxxx/' | go run main.go > mod.log
+```
+
+The string `EZE8HX` was present in my randomly generated file `data` only once and the `sed` command above changed it to the string `xxxxxx`. When we now compare the two log files using `diff`, we can see that for exactly one chunk the SHA-256 hash has changed, but the other chunks (and the chunk boundaries) stayed the same:
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: > but the other chunks (and the chunk boundaries) stayed the same:

Maybe you can omit the parenthesis here?

File: content/advent-2018/split-data-with-cdc.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+ 84093123       889631  0007ddacabc00000        a2b6e8651ae1ab69[...]
+ 84982754       2561935 000c53712f500000        bca589959b019a80[...]
+-87544689       2744485 0000a49118b00000        02b125104f06ed85[...]
++87544689       2744485 0000a49118b00000        ce2e43496ae6e7c3[...]
+ 90289174       1167308 00034d6cce700000        ad25a331993d9db7[...]
+ 91456482       1719951 001d2fea8ae00000        ba5153845c5b5228[...]
+ 93176433       1362655 0003009fc1600000        8e7a35e340c10d61[...]
+```
+
+This technique can be used for many other things besides a backup program.
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: It would be nice to have some examples here.

File: content/advent-2018/split-data-with-cdc.md [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+If you're interested in how the `chunker` package is used in restic, there's a [post over at the restic blog](https://restic.net/blog/2015-09-12/restic-foundation1-cdc) which explains this in a bit more detail.
+
+Performance
+===========
+
+Performance in a backup program is important: if it's too slow people tend to skip backup, which is undesirable. The `chunker` package is written in plain Go (so no fancy assembler functions or optimizations). I've optimized the calculations a bit and it's pretty fast already, the package contains some benchmarks:
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: > Performance in a backup program is important:

I would rephrase and split it into something like the following:

"Performance is crucial for backup programs. If creating backups is too slow people tend to stop using it.

`chunker` is written in plain Go. It doesn't have any fancy low-level assembler, but the code is already pretty fast, thanks to some optimized calculations. The package has some benchmarks:"

File: content/advent-2018/split-data-with-cdc.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+pkg: github.com/restic/chunker
+BenchmarkChunker-4            20          70342669 ns/op         477.01 MB/s
+--- BENCH: BenchmarkChunker-4
+    chunker_test.go:325: 23 chunks, average chunk size: 1458888 bytes
+    chunker_test.go:325: 23 chunks, average chunk size: 1458888 bytes
+PASS
+ok      github.com/restic/chunker       1.655s
+```
+
+You can see that data is processed at about 477 MB per second on a single core on an older machine using an Intel i5 CPU.
```
Comment (narqo) [inline content/advent-2018/split-data-with-cdc.md]: Maybe rephrase a bit:

"Running on an old machine with Intel i5 CPU `chunker` processes at about 477 MB per second on a single core."

File: content/advent-2018/split-data-with-cdc.md [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+
+```go
+p, err := chunker.RandomPolynomial()
+if err != nil {
+	panic(err)
+}
+
+fmt.Printf("using polynomial %v for splitting data\n", p)
+
+chunker := chunker.New(os.Stdin, p)
```
[the remark this answers is not in the document]
Comment (fd0) [reply content/advent-2018/split-data-with-cdc.md]: Good idea, let's use `chk`.

File: content/advent-2018/split-data-with-cdc.md (hunk 1 above)
Comment (fd0) [reply content/advent-2018/split-data-with-cdc.md]: Both good points, thanks for the ideas! And you're right, it's a *method* ;)

File: content/advent-2018/split-data-with-cdc.md (hunk 7 above)
Comment (fd0) [reply content/advent-2018/split-data-with-cdc.md]: I've added a few, good point!

File: content/advent-2018/split-data-with-cdc.md (hunk 8 above)
Comment (fd0) [reply content/advent-2018/split-data-with-cdc.md]: LGTM

File: content/advent-2018/split-data-with-cdc.md (hunk 9 above)
Comment (fd0) [reply content/advent-2018/split-data-with-cdc.md]: Ok

**Comment by fd0** [comment]
Thank you very much, I've integrated your comments!

**Comment by dgryski** [comment]
Thanks!
